### PR TITLE
AV-1831: E2E

### PR DESCRIPTION
**Pull request by dasconnor**
Adds an E2E test framework for executing tests against the compiled avalanche-cli binary. Tests exercise several commands and even execute hardhat solidity tests against a running EVM subnet.

### CODE REVIEW

**Comment by dasconnor** [comment]
This is ready for review now.

File: scripts/run.e2e.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+./tests/e2e/e2e.test --ginkgo.v
+
+EXIT_CODE=$?
+
+if [[ ${EXIT_CODE} -gt 0 ]]; then
+  echo "FAILURE with exit code ${EXIT_CODE}"
+  exit ${EXIT_CODE}
+else
+  echo "ALL SUCCESS!"
+fi
```
Comment (felipemadero) [inline scripts/run.e2e.sh:13]: minor detail for eol

File: tests/e2e/commands/subnet.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	// Config should no longer exist
+	exists, err = utils.SubnetConfigExists(subnetName)
+	gomega.Expect(err).Should(gomega.BeNil())
+	gomega.Expect(exists).Should(gomega.BeFalse())
+}
+
+// Returns the deploy output
+/* #nosec G204 */
+func DeploySubnetLocally(subnetName string) string {
+	// Check config does not already exist
```
Comment (felipemadero) [inline tests/e2e/commands/subnet.go:58]: really checking that exists

File: tests/e2e/commands/subnet.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	exists, err := utils.SubnetConfigExists(subnetName)
+	gomega.Expect(err).Should(gomega.BeNil())
+	gomega.Expect(exists).Should(gomega.BeTrue())
+
+	// Create config
```
Comment (felipemadero) [inline tests/e2e/commands/subnet.go:63]: this comment belongs to other call

File: tests/e2e/network/suite.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		deployOutput := commands.DeploySubnetLocally(subnetName)
+		_, err := utils.ParseRPCFromDeployOutput(deployOutput)
+		gomega.Expect(err).Should(gomega.BeNil())
+
+		commands.StopNetwork()
+		restartOutput := commands.StartNetwork()
+		rpc, err := utils.ParseRPCFromRestartOutput(restartOutput)
+		gomega.Expect(err).Should(gomega.BeNil())
+
+		err = utils.SetHardhatRPC(rpc)
```
Comment (felipemadero) [inline tests/e2e/network/suite.go:34]: maybe you can check some snapshot preservation, like rpc equalitty before/after

File: tests/e2e/utils/constants.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package utils
+
+const (
+	CLIBinary   = "./bin/avalanche"
```
Comment (felipemadero) [inline tests/e2e/utils/constants.go:4]: this seems duplicated with the other constants file. besides that, why not a unique constants for all e2e?

File: tests/e2e/utils/helpers.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"strings"
+
+	"github.com/ava-labs/avalanche-cli/pkg/constants"
+)
+
+func GetBaseDir() string {
```
Comment (felipemadero) [inline tests/e2e/utils/helpers.go:16]: don't we have this somewhere else? we should if not, probably

File: tests/e2e/commands/subnet.go (hunk 2 above)
Comment (dasconnor) [reply tests/e2e/commands/subnet.go]: fixed

File: tests/e2e/network/suite.go (hunk 4 above)
Comment (dasconnor) [reply tests/e2e/network/suite.go:60]: added a more sophisticated check

File: tests/e2e/utils/constants.go (hunk 5 above)
Comment (dasconnor) [reply tests/e2e/utils/constants.go]: I deleted the duplicate. I don't think there are a lot of wide spanning constants that are useful in multiple packages. Most of the use cases are contained, so I didn't break it out into its own package. We can add it later if we want.

File: tests/e2e/network/suite.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+		err := utils.DeleteConfigs(subnetName)
+		gomega.Expect(err).Should(gomega.BeNil())
+	})
+
+	ginkgo.It("can stop and restart a deployed subnet", func() {
+		genesis := "tests/e2e/genesis/test_genesis.json"
+
+		commands.CreateSubnetConfig(subnetName, genesis)
+		deployOutput := commands.DeploySubnetLocally(subnetName)
+		_, err := utils.ParseRPCFromDeployOutput(deployOutput)
```
[the remark this answers is not in the document]
Comment (dasconnor) [reply tests/e2e/network/suite.go]: I didn't explicitly add this check, but I did check the state instead.

**Review by holisticode — APPROVED** [review]

File: pkg/binutils/processes.go [hunk 8]
```diff
@@ -204,8 +205,10 @@ func KillgRPCServerProcess(app *app.Avalanche) error {
 		return fmt.Errorf("failed killing process with pid %d: %s", pid, err)
 	}
 
-	if err := os.Remove(app.GetRunFile()); err != nil {
-		return fmt.Errorf("failed removing run file %s: %s", constants.ServerRunFile, err)
+
```
[the remark this answers is not in the document]
Comment (dasconnor) [reply pkg/binutils/processes.go]: fixed

File: pkg/binutils/processes.go [hunk 9]
```diff
@@ -152,10 +154,9 @@ func StartServerProcess(app *app.Avalanche) error {
 	if err != nil {
 		return err
 	}
-	serverRunFile := path.Join(outputDir, constants.ServerRunFile)
-	app.SetRunFile(serverRunFile)
-	err = os.WriteFile(serverRunFile, rfBytes, perms.ReadWrite)
-	if err != nil {
+
+	serverRunFilePath := filepath.Join(app.GetRunDir(), constants.ServerRunFile)
```
[the remark this answers is not in the document]
Comment (dasconnor) [reply pkg/binutils/processes.go]: fixed

**Review by holisticode — APPROVED** [review]

**Pull request merged**